substratevm/src/com.oracle.svm.core.posix/src/com/oracle/svm/core/posix/JavaUtilZipSubstitutions.java

```diff
@@ -383,75 +504,145 @@ private int inflateBytes(long addr, byte[] b, int off, int len) throws DataForma
             } else if (ret == ZLib.Z_MEM_ERROR()) {
                 throw new OutOfMemoryError();
             } else {
-                throw new InternalError();
+                throw new InternalError(CTypeConversion.toJavaString(strm.msg()));
+            }
+        }
+    }
+
+    // (JDK 11)src/java.base/share/native/libzip/Inflater.c-static jlong doInflate(JNIEnv *env, jobject this, jlong addr,
+    //                       jbyte *input, jint inputLen,
+    //                       jbyte *output, jint outputLen)
+    private long doInflate(long addr, CCharPointer input, int inputLen, CCharPointer output, int outputLen) throws DataFormatException {
+        z_stream strm = WordFactory.pointer(addr);
+        int inputUsed = 0;
+        int outputUsed = 0;
+
+        strm.set_next_in(input);
+        strm.set_next_out(output);
+        strm.set_avail_in(inputLen);
+        strm.set_avail_out(outputLen);
+
+        int ret = ZLib.inflate(strm, ZLib.Z_PARTIAL_FLUSH());
+
+        int finished = 0;
+        int needDict = 0;
+
+        if (ret == ZLib.Z_STREAM_END() || ret == ZLib.Z_OK()) {
+            if (ret == ZLib.Z_STREAM_END()) {
+                finished = 1;
+                inputUsed = inputLen - strm.avail_in();
+                outputUsed = outputLen - strm.avail_out();
             }
+        } else if (ret == ZLib.Z_NEED_DICT()) {
+            needDict = 1;
+            inputUsed = inputLen - strm.avail_in();
+            outputUsed = outputLen - strm.avail_out();
+        } else if (ret == ZLib.Z_DATA_ERROR()) {
+            // these fields act as "out" parameters
+            inputConsumed = inputLen - strm.avail_in();
+            outputConsumed = outputLen - strm.avail_out();
+            throw new DataFormatException(CTypeConversion.toJavaString(strm.msg()));
+        } else if (ret == ZLib.Z_MEM_ERROR()) {
+            throw new OutOfMemoryError();
+        } else if (ret != ZLib.Z_BUF_ERROR()) {
+            throw new InternalError(CTypeConversion.toJavaString(strm.msg()));
```

Comment:
The C code is
```
    case Z_BUF_ERROR:
        break;
```
and then a return of OR of the status bits.  Rather than throwing an `InternalError`.

Reply:
But this is ``!=``, not ``==``.

Reply:
You are correct.